src/geometry/manifold/algebra/left_invariant_derivation.lean

```diff
@@ -0,0 +1,194 @@
+@[simp] lemma to_derivation_eq_coe : X.to_derivation = X := rfl
+
+lemma coe_injective :
+  @function.injective (left_invariant_derivation I G) (_ → C^⊤⟮I, G; 𝕜⟯) coe_fn :=
+λ X Y h, by { cases X, cases Y, congr', exact derivation.coe_injective h }
+
+@[ext] theorem ext (h : ∀ f, X f = Y f) : X = Y :=
+coe_injective $ funext h
+
+variables (X Y f)
+
+lemma coe_derivation :
+  ⇑(X : derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯) = (X : C^∞⟮I, G; 𝕜⟯ → C^∞⟮I, G; 𝕜⟯) := rfl
+
+lemma coe_derivation_injective : function.injective
+  (coe : left_invariant_derivation I G → derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯) :=
+λ X Y h, by { cases X, cases Y, congr, exact h }
+
+/-- Premature version of the lemma. Prefer using `left_invariant` instead. -/
+lemma left_invariant' :
+  (𝒅(𝑳 I g)) (1 : G) (derivation.eval_at (1 : G) ↑X) f = derivation.eval_at g ↑X f :=
+by rw [←to_derivation_eq_coe]; exact left_invariant'' X f g
+
+@[simp] lemma map_add : X (f + f') = X f + X f' := derivation.map_add X f f'
+@[simp] lemma map_zero : X 0 = 0 := derivation.map_zero X
+@[simp] lemma map_neg : X (-f) = -X f := derivation.map_neg X f
+@[simp] lemma map_sub : X (f - f') = X f - X f' := derivation.map_sub X f f'
+@[simp] lemma map_smul : X (r • f) = r • X f := derivation.map_smul X r f
+
+instance : has_zero (left_invariant_derivation I G) := ⟨⟨0, λ f g,
+  by simp only [linear_map.map_zero, derivation.coe_zero]⟩⟩
+
+instance : inhabited (left_invariant_derivation I G) := ⟨0⟩
+
+instance : has_add (left_invariant_derivation I G) :=
+{ add := λ X Y, ⟨X + Y, λ f g, by simp only [linear_map.map_add, derivation.coe_add,
+  left_invariant', pi.add_apply]⟩ }
+
+instance : has_neg (left_invariant_derivation I G) :=
+{ neg := λ X, ⟨-X, λ f g, by simp only [linear_map.map_neg, derivation.coe_neg, left_invariant',
+  pi.neg_apply]⟩ }
+
+instance : has_sub (left_invariant_derivation I G) :=
+{ sub := λ X Y, ⟨X - Y, λ f g, by simp only [linear_map.map_sub, derivation.coe_sub,
+  left_invariant', pi.sub_apply]⟩ }
+
+@[simp] lemma coe_add : ⇑(X + Y) = X + Y := rfl
+@[simp] lemma coe_zero : ⇑(0 : left_invariant_derivation I G) = 0 := rfl
+@[simp] lemma coe_neg : ⇑(-X) = -X := rfl
+@[simp] lemma coe_sub : ⇑(X - Y) = X - Y := rfl
+@[simp, norm_cast] lemma lift_add :
+  (↑(X + Y) : derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯) = X + Y := rfl
+@[simp, norm_cast] lemma lift_zero :
+  (↑(0 : left_invariant_derivation I G) : derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯) = 0 := rfl
+
+instance : add_comm_group (left_invariant_derivation I G) :=
+coe_injective.add_comm_group _ coe_zero coe_add coe_neg coe_sub
+
+instance : has_scalar 𝕜 (left_invariant_derivation I G) :=
+{ smul := λ r X, ⟨r • X, λ f g, by { simp only [derivation.Rsmul_apply, algebra.id.smul_eq_mul,
+            mul_eq_mul_left_iff, linear_map.map_smul, left_invariant'], left, refl }⟩ }
+
+variables (r X)
+
+@[simp] lemma coe_smul : ⇑(r • X) = r • X := rfl
+@[simp] lemma leibniz : X (f * f') = f • X f' + f' • X f := X.leibniz' _ _
```

Comment:
This one probably also belongs up on line 88 - these are properties of the  map itself, not of the `has_scalar` instance on line 118, and putting the lemmas before you define that structure makes that clear.